cis194/week7/stevemao/JoinList.hs

```diff
@@ -0,0 +1,53 @@
+{-# OPTIONS_GHC -Wall #-}
+module JoinList where
+import Sized
+
+data JoinList m a = Empty
+  | Single m a
+  | Append m (JoinList m a) (JoinList m a)
+    deriving (Eq, Show)
+
+(+++) :: Monoid m => JoinList m a -> JoinList m a -> JoinList m a
+Empty +++ a = a
+a +++ Empty = a
+a +++ b = Append (tag a <> tag b) a b
+
+tag :: Monoid m => JoinList m a -> m
+tag Empty = mempty
+tag (Single m _) = m
+tag (Append m _ _) = m
+
+getIntSize :: Sized a => a -> Int
+getIntSize = getSize . size
+
+indexJ :: (Sized b, Monoid b) =>
+  Int -> JoinList b a -> Maybe a
+indexJ _ Empty = Nothing
+indexJ i (Single s a)
+  | i == getIntSize s = Just a
+  | otherwise = Nothing
+indexJ i (Append s a b)
+  | i < getIntSize s = indexJ i a
+  | otherwise = indexJ i b
+
+dropJ :: (Sized b, Monoid b) =>
+  Int -> JoinList b a -> JoinList b a
+dropJ _ Empty = Empty
+dropJ i l@(Single s _)
+  | i < getIntSize s = l
+  | otherwise = Empty
+dropJ i (Append s a b)
```

Comment:
I'm trying to understand the logic here. 
Does `s` stand for the sum of the sizes for `a` and `b`? if so, does `i < getIntSize s` mean we are dropping anything up to the total size? and `i > getIntSize` means we are over-dropping?

Reply:
No, it may or may not over-dropping but I know that the first dropped element will be in `b`. But `b` is another 0 based index so I need to do `i - getSize a`.

Reply:
last dropped element*